### ISSUE TITLE
[BUG] <开启flash_attn后推理速度不但没加快反而更慢了>

### ISSUE THREAD

**Issue by MrJungle1** [QwenLM/Qwen#474]
### 是否已有关于该错误的issue或讨论？ | Is there an existing issue / discussion for this?

- [X] 我已经搜索过已有的issues和讨论 | I have searched the existing issues / discussions

### 该问题是否在FAQ中有解答？ | Is there an existing answer for this in FAQ?

- [X] 我已经搜索过FAQ | I have searched FAQ

### 当前行为 | Current Behavior

开启flash_attn后推理速度不但没加快反而更慢了
![image](https://github.com/QwenLM/Qwen/assets/43968255/0f3ee3fe-bcdf-426f-985e-b7667fd45f26)

### 期望行为 | Expected Behavior

use_flash_attn = True开启后推理相比于use_flash_attn = False更快

### 复现方法 | Steps To Reproduce

运行提供的profile.py文件
max_experiment_times = 1
context_length_per_experiment = 1024
generate_length_per_experiment = 1024
quant_type = "bf16"

Qwen-7B

### 运行环境 | Environment

```Markdown
- OS:CentOS 7
- Python:3.10
- Transformers:4.34.0/4.32.0
- PyTorch:2.1.0
- CUDA (`python -c 'import torch; print(torch.version.cuda)'`):11.8
```

**Comment by logicwong** [comment]
@MrJungle1 请问你是更新到了hf的最新代码吗？我们在最新代码里使用了torch2.0的scaled_dot_product_attention（关闭flash后会选择调用这个函数），它会根据卡型和输入的tensor自动选择背后的算子（可能是flash_attn，也可能是memory-efficient）。使用这个函数后在短序列上速度跟官方flashattention差不多.  https://huggingface.co/Qwen/Qwen-14B-Chat/blob/237cd704b81edf69b22e09b7b61d1b3a2bc55d53/modeling_qwen.py#L573-L575

**Comment by MrJungle1** [comment]
@logicwong 
![image](https://github.com/QwenLM/Qwen/assets/43968255/78f6607c-1e72-4aff-833d-720b9909f97c)
目前在这个commitx下，所以你意思这是正常的，use_flash_attn = True or Flase 推理速度差不多？

**Comment by logicwong** [comment]
@MrJungle1 切最新commit吧（前几天推的），旧commit flash-attn的速度有些问题。新commit我们对w & w/o flashattn的速度都进行了优化，优化后两者速度差不多，且都比之前快30%左右（单卡）

**Comment by MrJungle1** [comment]
@logicwong 
![image](https://github.com/QwenLM/Qwen/assets/43968255/c4f6cf11-b5fd-4a42-affa-bedb027380f1)
最新代码，还是不开更快，所以是不是没必要开flash_attn了

**Comment by logicwong** [comment]
ok，你这种情况确实是不用开了。

**Comment by logicwong** [comment]
@MrJungle1 Reame更新了速度对比 https://github.com/QwenLM/Qwen#inference-performance, 可以参考一下

**Comment by lissettecarlr** [comment]
的确是不开快，4090-24G单卡下： 
`
Average generate speed (tokens/s): 32.136035515385764
GPU Memory cost: 16.429929733276367GB
Experiment setting:
seed = 1024
max_experiment_times = 1
context_length_per_experiment = 1
generate_length_per_experiment = 2048
use_flash_attn = True
quant_type = bf16
`
关闭后
`
Average generate speed (tokens/s): 33.55059307823262
GPU Memory cost: 16.466691970825195GB
Experiment setting:
seed = 1024
max_experiment_times = 1
context_length_per_experiment = 1
generate_length_per_experiment = 2048
use_flash_attn = False
quant_type = bf16
`
然后是8192：
`
Average generate speed (tokens/s): 24.762144793048325
GPU Memory cost: 22.528968811035156GB
Experiment setting:
seed = 1024
max_experiment_times = 1
context_length_per_experiment = 1
generate_length_per_experiment = 8192
use_flash_attn = True
quant_type = bf16
`
关闭后
`
Average generate speed (tokens/s): 29.67292238380515
GPU Memory cost: 22.470723152160645GB
Experiment setting:
seed = 1024
max_experiment_times = 1
context_length_per_experiment = 1
generate_length_per_experiment = 8192
use_flash_attn = False
quant_type = bf16
`
